lib/parser/source/tree_rewriter.rb

```diff
@@ -381,21 +402,28 @@ def check_range_validity(range)
         range
       end
 
-      def enforce_policy(event)
-        return if @policy[event] == :accept
-        return unless (values = yield)
-        trigger_policy(event, **values)
+      EVENT_TO_POLICY = {
+        crossing_deletions:     :@crossing_deletions,
+        different_replacements: :@different_replacements,
+        swallowed_insertions:   :@swallowed_insertions,
+      }.freeze
+
+      def policy(event)
+        return :raise if event == :crossing_insertions
+
+        instance_variable_get(EVENT_TO_POLICY.fetch(event))
       end
 
       POLICY_TO_LEVEL = {warn: :warning, raise: :error}.freeze
       def trigger_policy(event, range: raise, conflict: nil, **arguments)
-        action = @policy[event] || :raise
+        action = policy(event)
         diag = Parser::Diagnostic.new(POLICY_TO_LEVEL[action], event, arguments, range)
-        @diagnostics.process(diag)
+        engine = @diagnostics || self.class.default_diagnostics
```

Comment:
could be just `diagnostic` (or `@diagnostic` if you move initialization back to constructor)